### DISCUSSION TITLE
bootstrapping against github server

### DISCUSSION THREAD

**Discussion by mrafaquat** [fluxcd/flux2#2339] [Q&A]
Dear team we are currently using lastest flux version in a kubernetes cluster against a repo sitting in a GithubServer.  The bootstrap process was done through the recommended approach of a PAT token.   I have asked by the team to move away from a PAT token, so i thought about using a deploy key.  I see the instructions are more specific to a Git Server as opposed to github server, can a deploy key be used instead of a PAT token when using a GitHub server ?  (i did try it but it failed)

**Comment by stefanprodan** [comment]
`flux bootstrap github` uses the PAT to add the generated deploy key to the repository. Without a PAT, the Flux CLI can't access the GitHub API. If you don't want to provider a PAT for the CLI actions, then you can use `flux bootstrap git` that uses SSH only.

**Reply by mrafaquat** [reply]
@stefanprodan thank you for your feedback.  I did try this.  For testingm i have a private repo which i have ssh access to and a created a kubernetes kind cluster,  when i execute the `flux bootstrap git` command i receive the following error 
```
► cloning branch "main" from Git repository "ssh://git@github.com/[REDACTED]"
✗ failed to clone repository: ssh: handshake failed: knownhosts: key mismatch
```

i then generated a key pair and passed that thropugh following the instructions but this also generates the same error,  Doing manual `git clones` works fine.  Have you come across ? Thanks again. your feedback is appreciated

**Reply by baflo** [reply]
@stefanprodan I used the `flux bootstrap git` command to not need the PAT. However, it still requests to set up a deploy key. I just want to have read-only access. Is there a documentation what the deploy key is required for?

```sh
>: ~/code$ flux bootstrap git --url ssh://git@github.com/<user>/repo.git --branch dev
► cloning branch "dev" from Git repository "ssh://git@github.com/<user>/repo.git"
✔ cloned repository
► generating component manifests
✔ generated component manifests
✔ component manifests are up to date
► installing components in "flux-system" namespace
✔ installed components
✔ reconciled components
► determining if source secret "flux-system/flux-system" exists
► generating source secret
✔ public key: ecdsa-sha2-nistp384 AAAAE2VjZHNhLXNoYTItbmlzdHAzODQAAAAIb....02cFyg8Z+l9fqOS/XNttJIrlMUMg==
? Please give the key access to your repository? [y/N] █

```

**Reply by kingdonb** [reply]
This isn't requesting a PAT, it's providing you with an SSH public key. This is a precaution to ensure that Flux users only bootstrap from a repo that they control. You can always add sources that are not private which you do not control using `flux create source git` later, but the bootstrap repo doesn't do this, because it needs write access during the bootstrap.

Flux bootstrap either can gain write access through an SSH key that you facilitate (as in the above snippet) or via PAT, which is safer because it only uses write access for a moment, then drops the privilege and downgrades itself to a read-only SSH key by default. But if you use `flux bootstrap git` there is no GitHub or GitLab API, so the only way is to provide a read-write SSH key. You can substitute it later for a read-only key, or you can substitute the whole GitRepository resource for one with an `https://` public git resource URL.

It is generally considered best-practice to use a private repo for your infrastructure, so some authentication is going to be required.
